File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/1_binary_search/binary_search.cpp
#include <iostream>
#include <cassert>
#include <vector>
using namespace std;

int binary_search(const vector<int> &a, int x)
{
    int left = 0, right = a.size() - 1;
    while(left <= right){
        int mid = left + (right - left) / 2;

        if(a[mid] == x)
            return mid;
        else if(a[mid] > x)
            right = mid - 1;
        else
            left = mid + 1;
    }
    return -1;
}

int linear_search(const vector<int> &a, int x)
{
    for(size_t i = 0; i < a.size(); ++i){
        if(a[i] == x) return i;
    }
    return -1;
}

int main(void)
{
    int n;
    cin >> n;
    
    vector<int> a(n);
    for(size_t i = 0; i < n; ++i){
        cin >> a[i];
    }
  
    int m;
    cin >> m;
  
    vector<int> b(m);
    for(size_t i = 0; i < m; ++i){
        cin >> b[i];
    }

    for(size_t i = 0; i < m; ++i){
        cout << binary_search(a, b[i]) << ' ';
    }

    return 0;
}

File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/1_binary_search/binary_search.py
def binary_search(a, x):
    left, right = 0, len(a) - 1
    while left <= right:
        mid = int(left + (right - left) / 2)

        if a[mid] == x:
            return mid
        elif a[mid] > x:
            right = mid - 1
        else:
            left = mid + 1

    return -1

def linear_search(a, x):
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1

if __name__ == '__main__':
    temp = list(map(int, input().split()))
    n = temp[0]
    data = temp[1:]

    temp = list(map(int, input().split()))
    m = temp[0]
    s_data = temp[1:]
    for x in s_data:
        print(binary_search(data, x), end = ' ')
